Measure text

tocsoft opened the pull request:
Added a new method to measure the text.

Fixes #42 

While I was in here I tried to refactor, what owns the output resolution, and how we pass rendering options in to try and simplify the API.

tocsoft committed:
measure text

hoist moved rendering options from methos to constructor
and add measure text.

tocsoft committed:
refactor to reduce mental load on switching between font and pixel positions.

Update IGlyphRasterizer to be responsible for the Resolution.

tocsoft committed:
Add some comments

tocsoft committed:
use const for resolution

vidstige reviewed (changes requested):
Thanks for your contribution. Some remarks

- It feels weird to have a function called `Render`but then sometimes it does not render. I think this can be done in some clearer way.
- Regarding units, the thought here is to only use Font Units, until the very last moment when it is sent to the rasterizer, which obviously needs pixels units. Basically, everything before the rasterizer should work only with Font Units. The reason is two fold. 1) Make things simpler to understand as different units are not needed to be juggled in different parts of code. 2) Enable integer only operations as far as possible.
- Remove all empty docstring. They are only valuable if they contain some valuable information. For example `Initializes a new instance of the Renderer class.` is not very helpful. I already knew this at constructor declaration - `Renderer()`.
- FontSize should perhaps be a parameter to the Render call instead of a `RenderOptions`. I love the idea if RenderOptions class, but I would like to use it for things that have sane defaults and you don't want to change too often.
- Unit tests - Awesome that you are adding tests in same pr as feature. Sign of a top 10% developer. :+1:
- ℹ️  Unit / integration tests. I like to split my tests up into unit tests and integration tests. For measure and the like I would like to set things up so that we do not need to load files to test these things. It's fine for now, as there is no such structure set up right now and it would be quite a large change.

Again, thanks for your contribution.

tocsoft committed:
renamed RenderOptions to TextOptions

renamed to the RenderOptions as they made more sense at a per render call set of options not class wide.

renamed the private Render method to ProcessText.

tocsoft commented:
I've updated based on your feedback... I renamed `RenderOptions` to `TextOptions` as I felt that made more sense when its being passed into `Render(..)` / `Measure(..)` methods.

I also renamed the private 'Render' method to `ProcessText` to be a bit more generic.

tocsoft committed:
fixed .csproj

tocsoft commented:
@vidstige I've had a pass of cleaning up the unit tests in relation to measuring strings, the changes are https://github.com/tocsoft/NRasterizer/commit/e0a14073851ba50841afc493c04ddd20a026a997 do you want me to add that changes to this PR or would you rather I sent you this over as a separate PR? after/if you accept this one.

vidstige commented on NOpenType/Renderer.cs:
```diff
@@ -179,23 +177,70 @@ private static Point<int> GetMidPoint(Point<int> v1, Point<int> v2)
+            int multiplyer = _rasterizer.Resolution / PointsPerInch;
+            int divisor = EmSquare.Size / options.FontSize;
+            int xx = (int)(x * divisor) / multiplyer;
+            int yy = (int)(y * divisor) / multiplyer;
```
what going on here? Aren't the name weird? You multiply with `divisor` and dividy by `multiplier`?

Also you're loosing bits when dividing before multiplying. Remake these so that no bits are lost. For five-star also make a unit test that capture these lost bits.

tocsoft replied:
They are called `multiplier` and `divisor` because that's what they are called lower down in `RenderGlyph(..)` / `ToPixelRasterizer` where, from my understanding, they are being used to convert the font sizes into pixel sizes.

Here we are doing the opposite trying to get the font size from the pixel ones, thus the inverted names... i'll give this some thought and see if I can't make this names clearer.

Refactoring into an iterator and / or refactoring `ToPixelRasterizer` a little might make this cleaner and clearer.

vidstige replied:
Ah, yeah, that makes sense. Let's keep the names then. Just reorder the operations real quick to avoid loosing bits and we're good to go. 

Yeah I think the iterator approach will be awesome. Less code and pretty clear, still fast and sweet. Should not be too much work either, let me know if you get stuck.

tocsoft replied:
which bits are we loosing exactly? 

This is explicitly doing the `x * divisor` before the `(result) / multiplyer`

or am I missing something?

vidstige replied:
Ah, let's say emsquare is 256 and fontsize is 192 and you want to render at x=55. For clarity lets say target resolution is 72 and PointsPerInch as well. You'll want to get `xx = 55 * 256 / 192 == 73`, but as the code stands you get `xx = 55 * (256 / 192) == 55`. Hope this makes sense? You can plug some similar numbers into unit test to also test this automatically.

Also, you can drop the cast to int.

tocsoft replied:
ah ok... totally missed those 2 line just above ... should be easy enough to fix.

vidstige commented on NOpenType/Renderer.cs:
```diff
@@ -179,23 +177,70 @@ private static Point<int> GetMidPoint(Point<int> v1, Point<int> v2)
+                if (xx > width)
+                {
+                    width = xx;
+                }
```
What's this if-statement? Capped with?

tocsoft replied:
This statement was actually written with me preempting newline support... should probably be removed from this version and added back in once code for newlines is being written.

vidstige replied:
I see. Shouldn't that be something like

```c#
if (xx > width)
{
    xx = 0;
    yy += _typeface.LineHeight;
}
```

Or something along those lines. But yeah, let's remove for now.

vidstige commented:
@tocsoft ah, back from vaccation. :-) Sorry for letting you hang. I guess it's easiest for you to pull these changes + the cleanup commit into a new branch and make a new pull request against master?

Also I feel we need a refactor. It feels to quirky to have the `bool renderGlyph` argument to `ProcessText`, but rather there should a function returning iterator that iterates over glyph coordinates. You see where I am going with this? Then the measure text and draw text functions could both use this iterator. Sample signature `IEnumerable<Point2D> GlyphPositions(Point2D position, string text, int size, TextOptions options = DefaultTextOptions)` or something like that.

vidstige commented:
Perhaps start with creating the new pr as this branch has merge conflicts? You can also merge and resolve conflicts in here if it easier for you :+1:

vidstige commented:
Let's get this baby in anyway :)

tocsoft commented:
superseded by #48
